src/Graphex.hs

```diff
@@ -20,61 +20,61 @@ import           Graphex.LookingGlass
 import           Graphex.Search              (bfsOn, findFirst, flood)
 
 -- | Convert a dependency file to a graph.
-depToGraph :: GraphDef -> Graph
+depToGraph :: GraphDef -> Graph Text
 depToGraph GraphDef{..} = Graph (links <> allNodes)
     where
         links = Map.fromListWith (<>) [ (name from, Set.singleton (name to)) | Edge{..} <- edges]
         allNodes = Map.fromList [(label, mempty) | Node{..} <- Map.elems nodes]
         name = (.label) . (nodes Map.!) :: Text -> Text
 
 -- | Convert a graph back to our dependency file format.
-graphToDep :: Graph -> GraphDef
+graphToDep :: Graph Text -> GraphDef
 graphToDep (Graph m) = GraphDef {
     title = "Internal Package Dependencies",
     edges = [ Edge { from = k, to = v } | (k, vs) <- Map.assocs m, v <- Set.toList vs ],
     nodes = Map.fromList [ (k, Node k Nothing) | k <- Map.keys m ]
     }
 
 -- | Reverse all the arrows in the graphs.
-reverseEdges :: Graph -> Graph
+reverseEdges :: Ord a => Graph a -> Graph a
 reverseEdges (Graph m) = Graph $ Map.fromListWith (<>) [ (v, Set.singleton k) | (k, vs) <- Map.assocs m, v <- Set.toList vs ] <> Map.fromSet (const mempty) (Map.keysSet m)
 
 -- | Find the direct list of things that are referencing this module.
-directDepsOn :: Graph -> Text -> Set Text
+directDepsOn :: Ord a => Graph a -> a -> Set a
 directDepsOn = flip (Map.findWithDefault mempty) . unGraph
 
 -- | Flood fill to find all transitive dependencies on a starting module, excluding the original key.
-allDepsOn :: Graph -> Text -> Set Text
+allDepsOn :: Ord a => Graph a -> a -> Set a
 allDepsOn = flood . directDepsOn
 
 -- | Find an example path between two modules.
 --
 -- This is a short path, but the important part is that it represents how connectivy works.
-why :: Graph -> Text -> Text -> [Text]
+why :: Ord a => Graph a -> a -> a -> [a]
 why m from to = fromMaybe [] $ findFirst bfsOn head (\ks -> (:ks) <$> (Set.toList . directDepsOn m . head) ks) [from] ((== to) . head)
 
 -- | Find all paths between two modules as a restricted graph of the intersection of
 -- reachable nodes from the start and to the end.
-allPathsTo :: Graph -> Text -> Text -> Graph
+allPathsTo :: (NFData a, Ord a) => Graph a -> a -> a -> Graph a
```

Comment:
The NFData constraints are a bit of a leak.

Reply:
Yeah that's a little unfortunate, but not a big deal.